operator/pkg/controlplane/etcd/util.go

```diff
@@ -0,0 +1,74 @@
+package etcd
+
+import (
+	"fmt"
+	operatorv1alpha1 "github.com/karmada-io/karmada/operator/pkg/apis/operator/v1alpha1"
+	"github.com/karmada-io/karmada/operator/pkg/constants"
+	"github.com/karmada-io/karmada/operator/pkg/util"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"strconv"
+	"strings"
+)
+
+// ConfigureClientCredentials configures etcd client credentials for Karmada core and aggregated API servers
+func ConfigureClientCredentials(apiServerDeployment *appsv1.Deployment, etcdCfg *operatorv1alpha1.Etcd, name, namespace string) error {
+	etcdClientServiceName := util.KarmadaEtcdClientName(name)
+	etcdCertSecretName := util.KarmadaCertSecretName(name)
+	// It's possible the API server is to be connected to an external etcd cluster or an in-cluster etcd cluster.
+	// As such, the etcd connection configuration is omitted from the default deployment template and must be configured explicitly.
+	// External and local etcd configurations are mutually exclusive.
+	if etcdCfg.Local != nil {
```

Comment:
Thanks for your thoughtful comments. I appreciate your philosophy for keepings things as simple as possible as I also think the same. In keeping consistent with that, I can provide some more context as to why I took this approach:
- Today the API server deployments are built from statically configured manifests that include the etcd config like so:
[args](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controlplane/apiserver/mainfests.go#L53)
[volume mount](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controlplane/apiserver/mainfests.go#L114)
[volume](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controlplane/apiserver/mainfests.go#L122)
Today, there's only support for local etcd, so that works fine, but now that we're adding support for external etcd, we need to reconfigure the etcd settings to account for both use cases. As an example, [this](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controlplane/apiserver/mainfests.go#L56) will no longer work. There's also non-standard data key names for the local etcd credentials which we may want to change in the future. Today, there are encoded  in the static manifest as [magic strings](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controlplane/apiserver/mainfests.go#L53) even though we use [constants](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/constants/constants.go#L90) when generating the secrets .

- There's inconsistencies in how the etcd connection is configured for the 3 API servers. For instance, the core API server pods get the dedicated etcd client credentials [volume attached](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controlplane/apiserver/mainfests.go#L126), but the search aggregated API server gets a completely different [volume attached](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controlplane/search/mainfests.go#L83) that's used to mount the etcd credentials.

To keep things simple and consistent, I just removed the etcd connection config (i.e., args, volume mount and volume configs) from the static manifests and centralized all the logic in one plane. Centralized  for both local and external etcd configuration and is consistent for all 3 API servers.

Reply:
I haven't looked into the implementation yet, I think the `SecretRef` is the reference for `karmada-operator` to get the necessary certifications, it just skips the process of generating certifications for ETCD clients, in other words, the referencing Secret is not the one mounted to Karmada components. 
After the `karmada-operatior` get the certifications, it will build an `internal` secret used to mount to Karmada components. 
That's my thoughts, let me know if we are on the same page.

Reply:
@chaosi-zju , @RainbowMango , thanks for all the comments. I appreciate the feedback and spent some time to try implementing the alternative approach suggested. 

Unfortunately, although those are good ideas, because of the current state of the code, using them will widen the scope of the changes beyond what is needed to provide a simple implement of  this feature that is minimally invasive and well integrated into the current state of the code, and will make it harder to reason about those changes. For instance, the idea of using the secret reference as the source for generating an internal secret that is mounted into the pod can actually work, but there are things that must be considered to make that work. Many questions will surface at this point. For instance, how will we keep things in sync? 

Today, one limitation in the operator is that [none of the resources it manages have owner references, nor does it watch for any resource  besides the `Karmada` type](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controller/karmada/controller.go#L148). Instead, it has a [deinit worflow](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/deinit.go#L56)  to do garbage collection. As such, the operator can not heal the cluster if I accidentally delete the deployment for one of the control plane component. 

Also, part of this conversation is regarding exactly what changes, if any, we should make to the [static templates](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controlplane/apiserver/mainfests.go#L21) for this feature. I'd love to hear your thoughts on this, but I think we'll probably have to eventually enhance the operator so that resources are created using strongly typed code  instead of using those static templates. This will help to avoid issues like the one that was fixed by this [PR](https://github.com/karmada-io/karmada/pull/5546). I also think that will help significantly with code locality and readability.

Thanks for letting me know about the effort to standardize the use of certs across all the Karmada components. I think that's a great idea. I also think the ideas for improvement I mentioned above are, at the very least, worth exploring as part of separate efforts, but for now, can we just move forward with this approach to keep these changes from getting out of scope?

Reply:
Another thing I think that's worth exploring while on the topic of etcd is whether the internal etcd support by the Karmada operator should be officially discouraged for production usage. For testing purposes, it's fine as it makes it easy to spin up a fully functional control plane, but for production usage, I don't think it's at the maturity level that one would expect for a production environment. Building a production-grade etcd operator is not a trivial undertaking and getting the internal etcd support in this operator to reach that maturity level will require quite an investment. I don't think it's worth the undertaking for a few reasons, including the fact that one operator to manage not only an etcd cluster, but also a Karmada control plane will have too broad of a scope. As I mentioned above, building an etcd operator alone requires a lot of investment. In fact, there's a new working group that was recently built for this undertaking. For reference:
- https://www.youtube.com/watch?v=yjaVpiZf3sA
- https://github.com/kubernetes/community/issues/7937
- https://github.com/etcd-io/etcd-operator

Would love to hear your thoughts on this.

Reply:
> Unfortunately, although those are good ideas, because of the current state of the code, using them will widen the scope of the changes beyond what is needed to provide a simple implement of this feature that is minimally invasive and well integrated into the current state of the code, and will make it harder to reason about those changes. For instance, the idea of using the secret reference as the source for generating an internal secret that is mounted into the pod can actually work, but there are things that must be considered to make that work. Many questions will surface at this point. For instance, how will we keep things in sync?

Sounds like a good point!
And thanks for the reminder, I didn't consider that.
This brings the concern that if we only read the content of the secret to generate a new secret, changes to the original secret will not take effect automatically. But you want it will take effect automatically, am I right? The scenario might be replacing expired certificates?

Reply:
@RainbowMango , thanks for the comments.

This implementation is simple and will just use the secret as the source of the etcd client cert volume.

I agree with you that the operator should be enhanced to add owner references to resources of interest so that it can enqueue reconciliation requests for changes to those resources. As of today, it [does not do that](https://github.com/karmada-io/karmada/blob/release-1.11/operator/pkg/controller/karmada/controller.go#L150) for any of the control plane components. I actually started investigating how to implement that enhancement, but then realized that incorporating those changes as part of this PR would introduce too many changes at once. I plan to submit another proposal for that enhancement as a follow up to this PR.

Any other comments?

Reply:
Refer to #5363.
@chaosi-zju is proposing to unify the name of secrets used by Karmada components, which would ease the maintenance effort, and also make it easier for users to use tools like CertManager to rotate expired certificates.

> This implementation is simple and will just use the secret as the source of the etcd client cert volume.

As a tradeoff, can we restrict the name of the secret holding the ETCD certificates? For example, if the `ExternalEtcd` is provided, then the SecretRef.name must be, for example, `etcd-client-cert-for-operator`.

Note: the format of the secret name is still under discussion.

Reply:
@RainbowMango , thanks for letting me know about that effort. I took a look and think it's a great step in the right direction. 

I do see that approach working. Cluster API actually uses a [similar](https://cluster-api.sigs.k8s.io/tasks/external-etcd#creating-the-required-secrets) approach.

For a `Karmada` instance, maybe we can use this convention:
`<karmada-instance-name>-etcd-client-cert`

Looks like that's the naming convention being proposed for the ongoing standardization effort you referenced. I think it's sensible.

One think to consider though is how we should go about validating this when provided.

What do you think?

Reply:
> For a Karmada instance, maybe we can use this convention:
<karmada-instance-name>-etcd-client-cert

I like it. for this secret, it could be `karmada-operator-etcd-client-cert.`  @chaosi-zju What do you think?

> One think to consider though is how we should go about validating this when provided.

Since there is no `webhook` for this `Karmada` resource, so I guess we can only validate it in the controller reconcile loop.

In the short term we can just emit an event if no conventional secret found. 
In the long term, we might need to extend the status of the `Karmada` API to present the conditions.

Reply:
Yeah. I think that convention is sensible and has the added bonus of being self-documenting. @chaosi-zju , I see you've been doing some great work on the standardization effort and would love to hear your thoughts. What do you think?

For now, I can make the changes and use a constant so that it can be easily changed if we decide to go with a different naming convention before those changes are merged.

Reply:
@RainbowMango , thanks for the  constructive feedback.  Pushed changes to address the comments. Please take a look.

Reply:
Sure. OK.

Reply:
> we are working together to do our best to explore the best results, I believe it is meaningful~

Thanks for all your hard work. I believe so too.